casper/src/main/scala/io/casperlabs/casper/equivocations/EquivocationDetector.scala

```diff
@@ -117,4 +146,117 @@ object EquivocationDetector {
 
   private def creatorJustificationHash(block: Block): Option[BlockHash] =
     ProtoUtil.creatorJustification(block.getHeader).map(_.latestBlockHash)
+
+  private def toposortJDagDesc[F[_]: Monad: Log](
+      dag: DagRepresentation[F],
+      block: Block
+  ): StreamT[F, BlockMetadata] = {
+    implicit val blockTopoOrdering: Ordering[BlockMetadata] = DagOperations.blockTopoOrderingDesc
+    DagOperations.bfToposortTraverseF(
+      List(BlockMetadata.fromBlock(block))
+    )(
+      _.justifications
+        .traverse(j => dag.lookup(j.latestBlockHash))
+        .map(_.flatten)
+    )
+  }
+
+  /**
+    * Find equivocating validators that a block can see based on its direct justifications
+    *
+    * This method is not used to detect new equivocations when receiving new blocks. It is used to help calculating main parent in the `forkchoice`.
+    *
+    * We use `bfToposortTraverseF` to traverse from `latestMessageHashes` down beyond the minimal rank
+    * of base block of equivocationRecords. Since we have already validated `validatorBlockSeqNum`
+    * equals 1 plus that of previous block created by the same validator, if we find a duplicated
+    * value, we know the validator has equivocated.
+    *
+    * @param dag the block dag
+    * @param justificationMsgHashes generate from direct justifications
+    * @param equivocationTracker local tracker of equivocations
+    * @tparam F effect type
+    * @return validators that can be seen equivocating from the view of latestMessages
+    */
+  def detectVisibleFromJustificationMsgHashes[F[_]: Monad](
```

Comment:
How about?
```suggestion
  def detectVisibleFromJustifications[F[_]: Monad](
```
We can omit `MsgHashes` because that's implicit.